src/parser/docparser.hpp

```diff
@@ -0,0 +1,75 @@
+#pragma once
+#include <nlohmann/json.hpp>
+#include <array>
+
+namespace docgen {
+
+// Tags indicating possible states of a DocParser object.
+enum class State {
+    read = 0,
+    slash,
+    single_line,
+    block,
+    last = block // update this variable if last enum changes
+};
+
+namespace details {
+
+// forward declaration
+// Parse function depending on a particular state.
+template <State state>
+inline const char *parse(const char *begin, 
+                         const char *end,
+                         nlohmann::json& json);
```

Comment:
I think `json` should be made a member and passed to constructor—it should not change between calls to `parse`.

Reply:
why should our parser own this json? I disagree.